### PR TITLE
Add Tracy packages

### PR DESCRIPTION
**Pull request by msimberg**
This adds two packages for Tracy (https://github.com/wolfpld/tracy):

- `tracy-client` is the client-side integration library required for instrumenting an application. This has no dependencies and can be used without building the full GUI. This can be built with CMake.
- `tracy` is the GUI profiler application, and the command-line capture tool. This has a lot of dependencies so it's nice to have this separately. In addition, the server can run on another node and may never be built together with the client library. This is built with make.

For `tracy-client` I've added all the CMake options from https://github.com/wolfpld/tracy/blob/8af7060a691a6500afe5a9bd44c7dab80bd4a281/CMakeLists.txt#L36-L57, including the trivial `TRACY_ENABLE` option. I've kept the negation in the variants for the options that are negated just to keep things synchronized, though I don't necessarily love the `TRACY_NO_X` options. The defaults are the same as in Tracy's `CMakeLists.txt`.

I've also added a `shared` variant to `glfw` because of an issue in Tracy/glfw/something else (https://github.com/wolfpld/tracy/issues/110). I'm a bit unsure who is causing problems there... `tracy` uses `glfw +shared` because of this.

If someone wants to try this out keep in mind that an issue for `mesa` (https://github.com/spack/spack/issues/29994) means that `tracy` needs to be installed with `tracy ^mesa ~llvm`.

### CODE REVIEW

**Comment by msimberg** [comment]
> Are you interested in adopting any of these package(s)?

Nope @spackbot, not interested...

File: var/spack/repos/builtin/packages/tracy-client/package.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+# Copyright 2021-2021 Lawrence Livermore National Security, LLC and other
```
Comment (tldahlgren) [inline var/spack/repos/builtin/packages/tracy-client/package.py]:
```suggestion
# Copyright 2021-2022 Lawrence Livermore National Security, LLC and other
```

Comment (tldahlgren) [reply var/spack/repos/builtin/packages/tracy-client/package.py]: You also need to address the style issues to get more tests to run.

```
var/spack/repos/builtin/packages/tracy-client/package.py:6: [F401] 'os' imported but unused
[147](https://github.com/spack/spack/runs/6372548239?check_suite_focus=true#step:6:147)
  var/spack/repos/builtin/packages/tracy-client/package.py:12: [E501] line too long (148 > 88 characters)
[148](https://github.com/spack/spack/runs/6372548239?check_suite_focus=true#step:6:148)
  var/spack/repos/builtin/packages/tracy-client/package.py:51: [F601] dictionary key 'no_frame_image' repeated with different values
[149](https://github.com/spack/spack/runs/6372548239?check_suite_focus=true#step:6:149)
  var/spack/repos/builtin/packages/tracy-client/package.py:55: [F601] dictionary key 'no_frame_image' repeated with different values
[150](https://github.com/spack/spack/runs/6372548239?check_suite_focus=true#step:6:150)
  var/spack/repos/builtin/packages/tracy/package.py:6: [F401] 'os' imported but unused
[151](https://github.com/spack/spack/runs/6372548239?check_suite_focus=true#step:6:151)
  var/spack/repos/builtin/packages/tracy/package.py:12: [E501] line too long (153 > 88 characters)
  ```

**Review by tldahlgren — COMMENTED** [review]
Confirmed sha256 and note that the same software/hash is used for both new packages.

**Comment by tldahlgren** [comment]
IMO There should be only one package for `tracy`.  The complexity comes in mostly because the two rely on different build systems.  Yes, the dependencies wouldn't be "pretty" but the combination would be only on the order of 100+ lines.

Why?  First, they are both based on the same software repository. Keeping them separate means both packages have to be updated EVERY TIME there is a new version as well as any change that affects both.

Based on your description, it sounds like the base code should be the server and the client library a variant (for when they are built together) that is disabled by default.  If someone can build the client without the server, then they could both be variants, with the client disabled by default.

@alalazo As someone working on the multiple build system support ATM, what are your thoughts about using a single versus two packages for the same code base?

**Comment by tldahlgren** [comment]
IF you haven't seen it already, there is documentation that might be helpful:

- Conditional dependencies: https://spack.readthedocs.io/en/latest/packaging_guide.html#conditional-dependencies
- Use of the `when` directive to aid dependencies and conflicts:  https://spack.readthedocs.io/en/latest/packaging_guide.html#common-when-constraints

**Comment by msimberg** [comment]
Thanks @tldahlgren for the comments. I'm definitely not set on having two separate packages. I think it makes (slightly) more sense to have separate packages which is why I started out with this, but see your point about duplication. My argument for separate packages is that I think it will be the normal use case to separately instrument code with `tracy-client` and then have one installation of the GUI tool (it doesn't have to be compiled with the same compilers and stack in general as the application). At least for me the most important criterion would be that the client can be compiled without the GUI. If they are to be a single package then I think the default would be to have the client functionality always enabled (since it requires no dependencies and is very quick to compile) and have the GUI through a variant that defaults to false.

That said I'd be interested in hearing a bit more about the multi-build-system option. Another alternative: Can I define a base class for two different packages (with `PackageBase` as base?) which would define the common metadata (urls, versions) and then derive from that for separate client and GUI packages?

**Comment by tldahlgren** [comment]
> Thanks @tldahlgren for the comments. I'm definitely not set on having two separate packages. I think it makes (slightly) more sense to have separate packages which is why I started out with this, but see your point about duplication. My argument for separate packages is that I think it will be the normal use case to separately instrument code with `tracy-client` and then have one installation of the GUI tool (it doesn't have to be compiled with the same compilers and stack in general as the application). At least for me the most important criterion would be that the client can be compiled without the GUI. If they are to be a single package then I think the default would be to have the client functionality always enabled (since it requires no dependencies and is very quick to compile) and have the GUI through a variant that defaults to false.
> 
> That said I'd be interested in hearing a bit more about the multi-build-system option. Another alternative: Can I define a base class for two different packages (with `PackageBase` as base?) which would define the common metadata (urls, versions) and then derive from that for separate client and GUI packages?

🤔 Spack *does* support inheritance so it is possible to define a base class that other classes inherit from.  We have examples in our test packages:

- https://github.com/spack/spack/blob/develop/var/spack/repos/builtin.mock/packages/simple-inheritance/package.py
- https://github.com/spack/spack/blob/develop/var/spack/repos/builtin.mock/packages/patch-inheritance/package.py and https://github.com/spack/spack/blob/develop/var/spack/repos/builtin.mock/packages/patch/package.py

I don't think we're doing that wrt. leveraging a single repository+commit and different build systems.

**Comment by alalazo** [comment]
@msimberg What would be the workflow for somebody who wants to use `tracy` with his own package? Asking to understand better what is the relation between these two packages and the software getting profiled.

**Comment by msimberg** [comment]
@alalazo take this with a small grain of salt as I'm still quite new to using tracy, but this is my experience from integrating it with pika (for which I'm intending to add a `tracy` variant once this is in).

Instrumenting an application only requires the client library (in the current state of the PR linking to a library provided by `tracy-client`) and adding some macros in the application. If you're interested in reading more section 2.1 in the manual talks about this (https://github.com/wolfpld/tracy/releases/latest/download/tracy.pdf). The manual even recommends using submodules/fetchcontent to get the client for integration in an application/library, but they recently added support for separately building the client and doing `find_package(TracyClient)` which is my goal with the `tracy-client` package here.

Building the server is section 2.3 in https://github.com/wolfpld/tracy/releases/latest/download/tracy.pdf. The manual says `Ideally, it would be best to use the same version of the Tracy profiler on both client and server`, but this means that the two have to be "compatible", not exactly the same version. What I have done myself so far is actually compile the server on my laptop, forward appropriate ports when sshing into Piz Daint, and then run an application instrumented with (only!) the client on Piz Daint. This is one reason for at least allowing to compile only the client. The second reason is that the server is relatively heavy to build with many dependencies, and since the server and client are separate processes they don't need to be compiled with the same compiler etc. If I were to use the server on the same node as the client I would still only compile a single version of the server and use clients compiled with any compiler version to connect to it.

TL;DR I don't care _too_ much about single package vs. multiple packages, the important thing is that the client can be built standalone (and if in one package I think only building the client should be the default). If you think a single package would be more idiomatic then I'm happy to go with that as well.

**Comment by tldahlgren** [comment]
This approach was discussed during an internal meeting earlier today and it was decided that having separate packages in this case is fine.  So I will approve and merge.  Thanks for your patience @msimberg

**Review by tldahlgren — APPROVED** [review]
Confirmed sha256.

**Comment by msimberg** [comment]
> This approach was discussed during an internal meeting earlier today and it was decided that having separate packages in this case is fine. So I will approve and merge. Thanks for your patience @msimberg

No worries at all, thanks for reviewing! I wasn't at all sure what I was doing was a good approach so I appreciate you having a look.
